pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/CastStep.java

```diff
@@ -21,25 +21,101 @@
 
 @Builder
 public class CastStep implements TransformStep {
-
   private final SchemaType keySchemaType;
   private final SchemaType valueSchemaType;
 
   @Override
   public void process(TransformContext transformContext) {
-    if (transformContext.getKeySchema() != null) {
-      Object outputKeyObject = transformContext.getKeyObject();
-      Schema<?> outputSchema = transformContext.getKeySchema();
-      if (keySchemaType == SchemaType.STRING) {
-        outputSchema = Schema.STRING;
-        outputKeyObject = outputKeyObject.toString();
-      }
-      transformContext.setKeySchema(outputSchema);
-      transformContext.setKeyObject(outputKeyObject);
+    if (transformContext.getKeySchema() != null
+        && keySchemaType != null
+        && transformContext.getKeySchema().getSchemaInfo().getType() != keySchemaType) {
+      Object value = convertValue(transformContext.getKeyObject(), keySchemaType);
+      transformContext.setKeySchema(toSchema(keySchemaType));
+      transformContext.setKeyObject(value);
+    }
+    if (valueSchemaType != null
+        && transformContext.getValueSchema().getSchemaInfo().getType() != valueSchemaType) {
+      Object value = convertValue(transformContext.getValueObject(), valueSchemaType);
+      transformContext.setValueSchema(toSchema(valueSchemaType));
+      transformContext.setValueObject(value);
+    }
+  }
+
+  private Object convertValue(Object originalValue, SchemaType schemaType) {
+    switch (schemaType) {
+      case BYTES:
```

Comment:
There is `BYTEBUFFER` as well if you are targeting the exhaustive list of schemas.

Reply:
Here it's a `SchemaType` for which there's only `SchemaType.BYTES`